sdk/nodejs/docker.ts

```diff
@@ -220,16 +220,51 @@ export function checkRepositoryUrl(repositoryUrl: string) {
 }
 
 async function buildAndPushImageWorkerAsync(
-    baseImageName: string,
+    imageName: string,
     pathOrBuild: string | pulumi.Unwrap<DockerBuild>,
-    repositoryUrl: string,
+    repositoryUrl: string | undefined,
     logResource: pulumi.Resource,
     connectToRegistry: (() => pulumi.Input<Registry>) | undefined,
     skipPush: boolean): Promise<string> {
 
-    checkRepositoryUrl(repositoryUrl);
+    // if we got an unknown repository url, just set to undefined for the remainder of
+    // processing. The rest of the code can handle that.
+    if (pulumi.containsUnknowns(repositoryUrl)) {
+        repositoryUrl = undefined;
+    }
 
-    const tag = utils.getImageNameAndTag(baseImageName).tag;
+    if (repositoryUrl) {
+        checkRepositoryUrl(repositoryUrl);
+    }
+
+    // First, login and pulling from docker if we can.
+    const cacheFrom = await loginAndPullFromCacheAsync(imageName, pathOrBuild, repositoryUrl, logResource, connectToRegistry);
+
+    // Then actually kick off the build.
+    logEphemeral("Starting docker build...", logResource);
+    const buildResult = await buildImageAsync(imageName, pathOrBuild, cacheFrom, logResource);
+    logEphemeral("Completed docker build", logResource);
+
+    // If we have no repository url, then we definitely can't push our build result. Same if
+    // we're in preview.
+    if (skipPush || !repositoryUrl || pulumi.runtime.isDryRun()) {
+        return imageName;
+    }
+
+    // Finally, if this a real update, push the built images to docker.
+    logEphemeral("Starting docker push...", logResource);
+    const result = await pushImageAsync(repositoryUrl, buildResult, logResource);
+    logEphemeral("Completed docker build", logResource);
+
+    return result;
+}
```

Comment:
tried to make it very clear in teh top level function what the overall big-steps are that we're performing.  The nuts-and-bolts logic is now moved into the helpers.